ReSharper.FSharp/src/FSharp.Psi.Features/src/Daemon/Stages/PipeChainTypeHintStage.fs

```diff
@@ -0,0 +1,153 @@
+﻿namespace JetBrains.ReSharper.Plugins.FSharp.Psi.Features.Daemon.Stages
+
+open System
+open System.Collections.Generic
+open JetBrains.Application.Settings
+open JetBrains.DocumentModel
+open JetBrains.ProjectModel
+open JetBrains.ReSharper.Daemon.Stages
+open JetBrains.ReSharper.Plugins.FSharp
+open JetBrains.ReSharper.Plugins.FSharp.Daemon.Stages.Tooltips
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Features.Daemon.Highlightings
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Features.Util
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Impl.Tree
+open JetBrains.ReSharper.Plugins.FSharp.Settings
+open JetBrains.ReSharper.Psi.Tree
+open JetBrains.ReSharper.Feature.Services.Daemon
+open JetBrains.ReSharper.Plugins.FSharp.Daemon.Cs.Stages
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Tree
+open FSharp.Compiler.SourceCodeServices
+open FSharp.Compiler.Layout
+open JetBrains.Util.Logging
+
+[<RequireQualifiedAccess>]
+[<Struct>]
+type SameLinePipeHints =
+    | Show
+    | Hide
+
+type PipeOperatorVisitor(fsFile: IFSharpFile, checkResults: FSharpCheckFileResults, sameLinePipeHints: SameLinePipeHints) =
+    inherit TreeNodeVisitor<List<FSharpIdentifierToken * ITreeNode>>()
+
+    let document = fsFile.GetSourceFile().Document
+
+    let showSameLineHints =
+        match sameLinePipeHints with
+        | SameLinePipeHints.Show -> true
+        | SameLinePipeHints.Hide -> false
+
+    let visitBinaryAppExpr binaryAppExpr (consumer: List<FSharpIdentifierToken * ITreeNode>) =
+        if not (FSharpExpressionUtil.isPredefinedInfixOpApp "|>" binaryAppExpr) then () else
+
+        let opExpr = binaryAppExpr.Operator
+        let exprToAdorn = binaryAppExpr.LeftArgument
+
+        let argCoords = document.GetCoordsByOffset(exprToAdorn.GetTreeEndOffset().Offset)
```

Comment:
`exprToAdorn` _might_ be `null` here if there are parse errors in the code (if we manage to add recovery for missing left arguments in binary op expressions). `GetTreeEndOffset` is an extension method and is ready for that and it'll return `InvalidOffset` (represented as `-1`) in such case, but `IDocument.GetCoordsByOffset` will then fail with `ArgumentOutOfRangeException`, since it expects non-negative offset.

I think it would just a bit better to check if the node is actually not `null` before getting the offset/coords. `opExpr` nullness is already checked in `isPredefinedInfixOpApp`.

There's also a few extension members like `ITreeNode.StartLine` in `PsiUtil` module, that could be used here instead of getting coords manually.